FirebaseCore/Tests/Unit/Swift/FirebaseAppTests.swift

```diff
@@ -0,0 +1,415 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import XCTest
+@testable import FirebaseCore
+
+struct TestApp {
+  static let androidClientID = "correct_android_client_id"
+  static let apiKey = "correct_api_key"
+  static let customizedAPIKey = "customized_api_key"
+  static let clientID = "correct_client_id"
+  static let trackingID = "correct_tracking_id"
+  static let gcmSenderID = "correct_gcm_sender_id"
+  static let googleAppID = "1:123:ios:123abc"
+  static let databaseURL = "https://abc-xyz-123.firebaseio.com"
+  static let storageBucket = "project-id-123.storage.firebase.com"
+  static let deepLinkURLScheme = "comgoogledeeplinkurl"
+  static let newDeepLinkURLScheme = "newdeeplinkurlfortest"
+  static let bundleID = "com.google.FirebaseSDKTests"
+  static let projectID = "abc-xyz-123"
+}
+
+let testAppName1 = "test_app_name_1"
+let testAppName2 = "test_app_name_2"
+
+let firebaseDefaultAppName = "__FIRAPP_DEFAULT"
+let googleAppIDKey = "FIRGoogleAppIDKey"
+let firebaseAppNameKey = "FIRAppNameKey"
+let firebaseAppIsDefaultAppKey = "FIRAppIsDefaultAppKey"
+
+class FirebaseAppTests: XCTestCase {
+  override func setUp() {
+    super.setUp()
+    FirebaseApp.resetApps()
+  }
+
+  override class func tearDown() {
+    super.tearDown()
+  }
+
+  func testSwiftFlagWithSwift() {
+    XCTAssertTrue(FirebaseApp.firebaseUserAgent().contains("swift"))
+  }
+
+  func testConfigure() {
+    let expectedUserInfo = expectedUserInfoForApp(named: firebaseDefaultAppName,
+                                                  isDefaultApp: true)
+    expectAppConfigurationNotification(with: expectedUserInfo)
+
+    let configurationAttempt = {
+      try ExceptionCatcher.catchException {
+        FirebaseApp.configure()
+      }
+    }
+    XCTAssertNoThrow(try configurationAttempt())
+
+    do {
+      let app = try XCTUnwrap(FirebaseApp.app(), "Failed to unwrap default app")
+      XCTAssertEqual(app.name, firebaseDefaultAppName)
+      XCTAssertEqual(app.options.clientID, TestApp.clientID)
+      XCTAssertTrue(FirebaseApp.allApps?.count == 1)
+
+    } catch {
+      XCTFail("The default app does not exist")
+    }
+
+    // MARK: Should we check the registered libraries instances available?
+
+    waitForExpectations()
+  }
+
+  func testConfigureDefaultAppTwice() {
+    FirebaseApp.configure()
+
+    let secondConfigurationAttempt = {
+      try ExceptionCatcher.catchException {
+        FirebaseApp.configure()
+      }
+    }
+
+    XCTAssertThrowsError(try secondConfigurationAttempt())
```

Comment:
This test shows how you could use the `ExceptionCatcher` class. Configuring a FirebaseApp twice throws an exception, and I couldn't test the Swift API for this because the thrown objc exception can't be handled by in swift. This class allows you to provide its `catchException` method with a closure that contains code that might throw an objc exception and then handle it sorta? elegantly in swift.